src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ThreadStaticsNode.cs

```diff
@@ -57,28 +71,85 @@ public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFacto
 
             result.Add(new DependencyListEntry(GetGCStaticEETypeNode(factory), "ThreadStatic MethodTable"));
 
-            if (factory.PreinitializationManager.HasEagerStaticConstructor(_type))
+            if (_type != null)
             {
-                result.Add(new DependencyListEntry(factory.EagerCctorIndirection(_type.GetStaticConstructor()), "Eager .cctor"));
-            }
 
-            ModuleUseBasedDependencyAlgorithm.AddDependenciesDueToModuleUse(ref result, factory, _type.Module);
+                if (factory.PreinitializationManager.HasEagerStaticConstructor(_type))
+                {
+                    result.Add(new DependencyListEntry(factory.EagerCctorIndirection(_type.GetStaticConstructor()), "Eager .cctor"));
+                }
+
+                ModuleUseBasedDependencyAlgorithm.AddDependenciesDueToModuleUse(ref result, factory, _type.Module);
+            }
+            else
+            {
+                foreach (var type in _inlined.GetTypes())
+                {
+                    if (factory.PreinitializationManager.HasEagerStaticConstructor(type))
+                    {
+                        result.Add(new DependencyListEntry(factory.EagerCctorIndirection(type.GetStaticConstructor()), "Eager .cctor"));
+                    }
+
+                    ModuleUseBasedDependencyAlgorithm.AddDependenciesDueToModuleUse(ref result, factory, type.Module);
+                }
+            }
 
             return result;
         }
 
-        public override bool HasConditionalStaticDependencies => _type.ConvertToCanonForm(CanonicalFormKind.Specific) != _type;
+        public override bool HasConditionalStaticDependencies
+        {
+            get
+            {
+                if (_type != null)
+                {
+                    return _type.ConvertToCanonForm(CanonicalFormKind.Specific) != _type;
+                }
+                else
+                {
+                    foreach (var type in _inlined.GetTypes())
+                    {
+                        if (type.ConvertToCanonForm(CanonicalFormKind.Specific) != type)
```

Comment:
This condition basically checks whether this is shared generic type. This should not be needed given the current impl - the current impl does not work for shared generic types.